core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlock.java

```diff
@@ -158,28 +159,42 @@ public void retainedBytesForEachPart(BiConsumer<Object, Long> consumer)
     public Block copyPositions(int[] positions, int offset, int length)
     {
         checkArrayRange(positions, offset, length);
+        if (length == 0) {
+            return new VariableWidthBlock(0, 0, EMPTY_SLICE, new int[1], null);
+        }
 
+        int[] newOffsets = new int[length + 1];
         int finalLength = 0;
-        for (int i = offset; i < offset + length; i++) {
-            finalLength += getSliceLength(positions[i]);
+        for (int i = 0; i < length; i++) {
+            int position = positions[offset + i];
+            finalLength += getSliceLength(position);
+            newOffsets[i + 1] = finalLength;
         }
+
         SliceOutput newSlice = Slices.allocate(finalLength).getOutput();
-        int[] newOffsets = new int[length + 1];
         boolean[] newValueIsNull = null;
+        int firstPosition = positions[offset];
         if (valueIsNull != null) {
             newValueIsNull = new boolean[length];
+            newValueIsNull[0] = valueIsNull[firstPosition + arrayOffset];
         }
-
-        for (int i = 0; i < length; i++) {
+        int currentStart = getPositionOffset(firstPosition);
+        int currentEnd = getPositionOffset(firstPosition + 1);
+        for (int i = 1; i < length; i++) {
             int position = positions[offset + i];
-            if (!isEntryNull(position)) {
-                newSlice.writeBytes(slice, getPositionOffset(position), getSliceLength(position));
+            int currentOffset = getPositionOffset(position);
```

Comment:
Add a comment why this works for null positions.
I would still make null handling first so it's less confusing for future readers

Reply:
updated